content/advent-2019/flags.md

```diff
@@ -0,0 +1,279 @@
+	}
+
+	var err error
+	switch os.Args[1] { // [3]
+	case "run":
+		err = runHTTPD()
+	case "check":
+		err = checkHTTPD()
+	default:
+		err = fmt.Errorf("error: unknown command - %s", os.Args[1])
+	}
+
+	if err != nil {
+		log.Fatalf("error: %s", err)
+	}
+
+}
+
+func checkHTTPD() error {
+	fs := flag.NewFlagSet("check", flag.ContinueOnError) // [4]
+	fs.Usage = func() {
+		fmt.Fprintf(flag.CommandLine.Output(), checkUsage, os.Args[0])
+		fs.PrintDefaults()
+	}
+
+	if err := fs.Parse(os.Args[2:]); err != nil { // [5]
+		return err
+	}
+
+	if fs.NArg() != 1 {
+		return fmt.Errorf("error: wrong number of arguments")
+	}
+
+	url := fs.Arg(0) // [6]
+	resp, err := http.Get(url)
+	switch {
+	case err != nil:
+		return err
+	case resp.StatusCode != http.StatusOK:
+		return fmt.Errorf("error: bad status - %s", resp.Status)
+	}
+
+	return nil
+}
+
+func runHTTPD() error {
+	fs := flag.NewFlagSet("check", flag.ContinueOnError)
+	fs.Var(PortVar(&config.port), "port", "port to listen on")
+	fs.StringVar(&config.host, "host", config.host, "host to listen on")
+	fs.Usage = func() {
+		fmt.Fprintf(flag.CommandLine.Output(), httpdUsage, os.Args[0])
+		fs.PrintDefaults()
+	}
+
+	if err := fs.Parse(os.Args[2:]); err != nil {
+		return err
+	}
+
+	http.HandleFunc("/", handler)
+	addr := fmt.Sprintf("%s:%d", config.host, config.port)
+	fmt.Printf("server ready on %s\n", addr)
+	return http.ListenAndServe(addr, nil)
+}
+
+```
+
+1. Usage for the main executable
+2. We should have at lest two `os.Args` - the executable and the sub command name
+3. `os.Args[1]` is the subcommand (`os.Args[1]` is the executable name)
+4. Create a new `FlagSet` to parse the command line for this sub command. Use
+   `flag.ContinueOnError` so parse error will not exit the program. The only function
+   that should exit the program is `main`, all others should return an error
+5. Pass rest of arguments. e.g. `["app", "check", "http://localhost:8080"]` →
+   `["localhost:8081]`
+6. [fs.Arg](https://golang.org/pkg/flag/#FlagSet.Arg) returns the nth command
+   line argument (not including the program name) after parsing
+
+
+
+# Conclusion
+
+The `flag` package is flexible and will probably support all of your command
+line parsing needs. It might be more verbose than other packages but it's in
+the standard library so you don't need any extra dependencies and can count on
+its API keeping the [Go compatibility
+promise](https://golang.org/doc/go1compat).
+
+You can see the full source code for the examples
+[here](https://github.com/gopheracademy/gopheracademy-web/blob/master/content/advent-2019/flag).
+
+
+# About the Author
+Hi there, I'm Miki, nice to e-meet you ☺. I've been a long time developer and
+have been working with Go for about 10 years. I write code professionally as
+a consultant and contribute a lot to open source. Apart from that I'm a [book
+author](https://www.amazon.com/Forging-Python-practices-lessons-developing-ebook/dp/B07C1SH5MP) author, an author on [LinkedIn
```

Comment:
double `author`

Reply:
Thanks. You're the first out of 3 reviewers to catch this.